azure_functions_worker/functions.py

```diff
@@ -243,21 +230,70 @@ def add_function(self, function_id: str,
                     f'Python return annotation "{return_pytype.__name__}" '
-                    f'does not match binding type "{return_binding_name}"')
+                    f'does not match binding type "{binding_name}"')
 
         if has_implicit_return and 'return' in annotations:
             return_pytype = annotations.get('return')
 
         return_type = None
         if has_explicit_return or has_implicit_return:
-            return_type = ParamTypeInfo(return_binding_name, return_pytype)
+            return_type = ParamTypeInfo(binding_name, return_pytype)
+
+        return return_type
+
+    def add_function(self, function_id: str,
+                     func: typing.Callable,
+                     metadata: protos.RpcFunctionMetadata):
+        func_name = metadata.name
+        sig = inspect.signature(func)
+        params = dict(sig.parameters)
+        annotations = typing.get_type_hints(func)
+        return_binding_name: typing.Optional[str] = None
+        has_explicit_return = False
+        has_implicit_return = False
+
+        bound_params = {}
+        for binding_name, binding_info in func.metadata.bindings.items():
+
+            self.validate_binding_direction(binding_name,
+                                            binding_info.direction, func_name)
+
+            if binding_name == '$return':
+                return_binding_name = binding_info.type
+                has_explicit_return = True
+                assert binding_name is not None
+            elif bindings_utils.has_implicit_output(binding_info.type):
+                # If the binding specify implicit output binding
+                # (e.g. orchestrationTrigger, activityTrigger)
+                # we should enable output even if $return is not specified
+                return_binding_name = binding_info.type
+                has_implicit_return = True
+                bound_params[binding_name] = binding_info
+            else:
+                bound_params[binding_name] = binding_info
+
+        requires_context = self.is_context_required(params, bound_params,
+                                                    annotations,
+                                                    func_name)
+
+        input_types, output_types = self.validate_function_params(params,
+                                                                  bound_params,
+                                                                  annotations,
+                                                                  func_name)
+
+        return_type = \
+            self.validate_annotations_and_return_type(annotations,
+                                                      has_explicit_return,
+                                                      has_implicit_return,
+                                                      return_binding_name,
+                                                      func_name)
 
         self._functions[function_id] = FunctionInfo(
```

Comment:
I would even recommend creating an abstraction of "add_function_to_database" such that in the future if there is any optimization we need to do, or change the `FunctionInfo` data structure to simplify things (or make it more complex), that can be done in an easier fashion.